_posts/2020-06-01-rethinking-the-frontend.md

```diff
@@ -0,0 +1,227 @@
+(def counter (atom 0))
+
+;; function that takes in counter atom and renders the UI
+(defn counter-app [counter]
+  (h/html
+    [:div
+      "Counter value: " [::h/live {:source counter
+                                   :component #(h/out! (str %))}]
+      [:button {:on-click #(swap! counter inc)} "increment"]
+      [:button {:on-click #(swap! counter dec)} "decrement"]]))
+```
+
+The above example looks similar to [Reagent](https://github.com/reagent-project/reagent)
+(a ClojureScript React wrapper), but it is all rendered on the server.
+The button callbacks defined with `:on-click` are actually run on the server.
+The special `[::h/live ...]` element registers a source on the server for this
+page. Whenever the atom changes, Ripley re-renders the component on the server and sends
+the resulting HTML sent via WebSocket to the client.
+
+Ripley includes a tiny JS client library that does the WebSocket handling: sending
+event handler callbacks to the server and patching in new fragments received from
+the server.
+
+Ripley tries to be as efficient as possible and the HTML output is optimized at compile
+time by prerendering all static parts to HTML that can be directly written to the result
+stream.
+
+<!---
+@startuml
+group Initial page render
+  frontend -> backend: request index page
+  backend -> db: SQL query data
+  db -> backend: result rows
+  backend -> frontend: full index page
+  note left: browser renders the page\ncontaining the data\nuser wants to see
+end
+group Interaction after page has been loaded
+ note left of frontend: user clicks\nnext page
+ frontend -> backend: WS 1:[]
+ backend -> db: SQL query
+ db -> backend: result rows
+ backend -> frontend: WS 0:R:...new HTML content...
+ note left of frontend: frontend patches in\nupdated content
+end
+@enduml
+-->
+![ripley sequence](/img/2020-rethinking-frontend/ripley-sequence.png)
+
+The image above shows the same hypothetical application with Ripley. The client
+requests the page and directly receives HTML having all the content in place.
+The subsequent interaction is interesting as well. When user clicks the "next page"
+button, only the callback id and possible arguments are sent to the server via the
+WebSocket. The server will run the query and send only the changed part of the
+page back over the WebSocket.
+
+The basic model is that a live component is rerendered on the server and its full
+HTML is sent back but Ripley can also delete elements, append or prepend content
+or change attributes. The update granularity can be decided per component when
+implementing live components.
+
+
+The model has many implications and some trade-offs we will explore next.
+
+### Time to interactive
+
+SPA frontend apps are often large applications that the client needs to download before
+it can render the page for the user to see and interact with. This problem has some work
+arounds like module splitting and server side rendering with hydration. These solutions
```

Comment:
```suggestion
it can render the page for the user to see and interact with. This problem has some workarounds like module splitting and server side rendering with hydration. These solutions
```